### PR TITLE
embedding fraction implementation

### PR DESCRIPTION
**Pull request by samhavens**
Implement embedding fraction per CogView and GLM-130B. To use, model config should look like (last line is new, the rest is a normal config):

```yaml
# Model
model:
  name: mosaic_gpt
  device: meta
  tokenizer_name: *tokenizer_name
  d_model: 2048
  n_heads: 16 # Modified 24->16 so that d_head == 128 to statisfy FlashAttention
  n_layers: 24
  mlp_ratio: 4
  max_seq_len: *max_seq_len
  vocab_size: 50257
  init_std: 0.02
  attn_pdrop: 0.0
  resid_pdrop: 0.0
  emb_pdrop: 0.0
  attn_impl: flash
  embedding_fraction: 0.1  # This is the value GLM-130B used
```

I noticed that a lot of config values were mandatory, but I made this default to 1 since it is not common.

### CODE REVIEW

**Review by jemdohmann — APPROVED** [review]
Thank you Sam! have you run this yet? If not I will run a 760m param run to add to my loss spikes experiment

**Comment by samhavens** [comment]
I haven't tried it yet! Sorry 😅  should have noted that

File: llm/src/mosaic_gpt.py [hunk 1]
```diff
@@ -186,7 +188,13 @@ def forward(self,
 
         tok_emb = self.transformer.wte(input_ids)  # type: ignore
         pos_emb = self.transformer.wpe(pos)  # type: ignore
-        x = self.transformer.emb_drop(tok_emb + pos_emb)  # type: ignore
+        if self.embedding_fraction == 1:
+            x = self.transformer.emb_drop(tok_emb + pos_emb)  # type: ignore
+        else:
+            x = tok_emb + pos_emb
+            x = self.transformer.emb_drop(
+                x = x * self.embedding_fraction + x.detach() * (1 - self.embedding_fraction)
```
Comment (dakinggg) [inline llm/src/mosaic_gpt.py:196]: Is the variable assignment in the expression a typo?

File: llm/src/mosaic_gpt.py [hunk 2]
```diff
@@ -145,6 +145,8 @@ def __init__(self, cfg: DictConfig):
         super().__init__()
         assert cfg.name == 'mosaic_gpt', f'Tried to build MosaicGPT model with cfg.name={cfg.name}'
         self.cfg = cfg
+        # CogView and GLM-130B papers both report this helping with stabilizing training
+        self.embedding_fraction = cfg.embedding_fraction if 0 < cfg.embedding_fraction < 1 else 1
```
[the remark this answers is not in the document]
Comment (samhavens) [reply llm/src/mosaic_gpt.py]: yes, done

File: llm/src/mosaic_gpt.py (hunk 1 above)
Comment (samhavens) [reply llm/src/mosaic_gpt.py]: yup. fixed

**Review by dakinggg — APPROVED** [review]
LGTM, although it would be great if you could add a link to where the implementation comes from

**Comment by samhavens** [comment]
> LGTM, although it would be great if you could add a link to where the implementation comes from

I added a links to the paper where this comes from (CogView) and the paper which suggests this implementation

**Comment by dblalock** [comment]
This looks small and approved so I'm gonna merge it.
